### PR TITLE
Add Statement and Database\Query::all()

### PR DESCRIPTION
**Pull request by othercorey**
This has several key changes:

* Removes support for callbacks in statements.
* Remove nested statement design in favor of single implementation.
* Move field type conversion directly into statement instead of using callbacks.
* Add Database\Query::all() which returns a result set for SELECT statements.

Users would be required to call `all()` for result set decorators to run. 

This enforces result set modification is done outside of the Statement which should be a straight-forward wrapper around `PDOStatement`.

Key issues;

* ~`Database\Query` owns result decorators which used to run via callbacks, but now must be wrapped in a result set separate from `execute()`.~
* ~`Database\Query::all()` returns `iterable` instead of `array` to allow `Datasource\QueryTrait::all()` to override it with `ResultSetInterface`.~
* `QueryTrait` declares `$_results` but it's already declared in `Database\Query`. This is because `ORM\Query` is implemented by a Datasource trait.
* `SqliteStatement` is still guessing what type of query was run to perform `SELECT CHANGES()`.

### CODE REVIEW

File: src/ORM/Query.php [hunk 1]
```diff
@@ -1126,7 +1125,8 @@ protected function _execute(): iterable
             return $this->_results;
         }
 
-        $results = $this->execute()->fetchAll(StatementInterface::FETCH_TYPE_ASSOC);
+        /** @var array $results */
```
Comment (othercorey) [inline src/ORM/Query.php]: This is a hack since we know Database\Query actually uses an array.

File: src/Database/Query.php [hunk 2]
```diff
@@ -237,15 +242,40 @@ public function getConnection(): Connection
...
+    public function execute(): Statement
     {
-        $statement = $this->_connection->run($this);
-        $this->_iterator = $this->_decorateStatement($statement);
+        $this->_statement = $this->_connection->run($this);
         $this->_dirty = false;
 
-        return $this->_iterator;
+        return $this->_statement;
+    }
```
Comment (othercorey) [inline src/Database/Query.php:274]: This should throw an exception if result decorators were added, but I'm currently calling `execute()` from `all()`.

File: src/Database/Statement.php [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+namespace Cake\Database;
+
+use Cake\Database\Log\LoggedQuery;
+use InvalidArgumentException;
+use PDO;
+use PDOException;
+use PDOStatement;
+use Psr\Log\LoggerInterface;
+
+class Statement
```
Comment (ADmad) [inline src/Database/Statement.php]: Why did you remove the interface?

File: src/Database/Driver.php [hunk 4]
```diff
@@ -198,12 +202,18 @@ abstract public function enabled(): bool;
...
      */
-    public function prepare(Query|string $query): StatementInterface
+    public function prepare(Query|string $query): Statement
     {
         $this->connect();
         $statement = $this->_connection->prepare($query instanceof Query ? $query->sql() : $query);
 
-        return new PDOStatement($statement, $this);
+        $typeMap = null;
+        if ($query instanceof Query && $query->isResultsCastingEnabled() && $query->type() === 'select') {
```
Comment (ADmad) [inline src/Database/Driver.php]: Would be nice to have constants for query types instead of using strings.

File: src/Database/Driver/Sqlserver.php [hunk 5]
```diff
@@ -212,7 +209,12 @@ public function prepare(Query|string $query): StatementInterface
             ]
         );
 
-        return new SqlserverStatement($statement, $this);
+        $typeMap = null;
+        if ($query instanceof Query && $query->isResultsCastingEnabled() && $query->type() === 'select') {
+            $typeMap = $query->getSelectTypeMap();
+        }
+
+        return new SqlserverStatement($this, $statement, $typeMap);
```
Comment (ADmad) [inline src/Database/Driver/Sqlserver.php]: The `STATEMENT_CLASS` constant should be used.

File: src/Database/Query.php [hunk 6]
```diff
@@ -2395,7 +2404,7 @@ function ($errno, $errstr): void {
             'params' => $params,
             'defaultTypes' => $this->getDefaultTypes(),
             'decorators' => count($this->_resultDecorators),
-            'executed' => $this->_iterator ? true : false,
+            'executed' => $this->_statement !== null ? true : false,
```
Comment (ADmad) [inline src/Database/Query.php]:
```suggestion
            'executed' => (bool)$this->_statement,
```

File: src/Database/Statement.php [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+     * take an array or arguments to be bound to the query variables. Please note
+     * that binding parameters from this method will not perform any custom type conversion
+     * as it would normally happen when calling `bindValue`.
+     *
+     * @param array|null $params list of values to be bound to query
+     * @return bool true on success, false otherwise
+     */
+    public function execute(?array $params = null): bool
+    {
+        $loggedQuery = new LoggedQuery();
```
Comment (ADmad) [inline src/Database/Statement.php]: Shouldn't the `LoggedQuery` instance be created only if logging is enabled?

File: src/Database/Statement.php [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+     * @throws \InvalidArgumentException
+     */
+    protected function convertMode(string|int $mode): int
+    {
+        if (is_int($mode)) {
+            // We don't try to validate the PDO constants
+            return $mode;
+        }
+
+        static $MODES = ['assoc' => PDO::FETCH_ASSOC, 'num' => PDO::FETCH_NUM, 'obj' => PDO::FETCH_OBJ];
```
Comment (ADmad) [inline src/Database/Statement.php]: Why not use a protected static class property?

File: src/Database/Statement.php [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+     * Returns the number of rows affected by the last SQL statement.
+     *
+     * This behaves the same as `PDOStatement::rowCount()`.
+     *
+     * @return int
+     * @link https://www.php.net/manual/en/pdostatement.rowcount.php
+     */
+    public function rowCount(): int
+    {
+        if ($this->typeConverter !== null) {
```
Comment (ADmad) [inline src/Database/Statement.php]: So this will effectively prevent use of `rowCount()` for SELECT queries even for dbs that support it isn't it? Not sure if we should do that. It could still be useful for debugging.

Comment (othercorey) [reply src/Database/Statement.php]: I don't believe it's ever supported that way, but we can do that.

File: src/Database/Statement.php (hunk 3 above)
Comment (othercorey) [reply src/Database/Statement.php]: Without the nested design, there is no need for an interface. We expect all instances to extend `Statement`.

File: src/Database/Driver.php (hunk 4 above)
Comment (othercorey) [reply src/Database/Driver.php]: I can add those back.

File: src/Database/Statement.php (hunk 7 above)
Comment (othercorey) [reply src/Database/Statement.php]: I would like to clean up this function, but I didn't figure anything out yet. I do want to avoid a bunch of if checks and temp variables. The requirement to capture and log the error makes things a little clunky.

File: src/Database/Statement.php (hunk 8 above)
Comment (othercorey) [reply src/Database/Statement.php]: I did originally. Tried it this way to see if it was cleaner.

**Comment by othercorey** [comment]
I removed ``lastInsertId()` because of the support it implies.

If the driver does not return inserted IDs as rows, then it calls `PDO::lastInsertId()` which does not operate on a specific statement or column and for some drivers not on a specific table (postgresqL

However, for drivers like MySQL, having to find the driver each time you want the auto-increment id, that might be tedious. This is hidden from the user for ORM inserts.

File: src/Database/Statement.php (hunk 3 above)
Comment (ADmad) [reply src/Database/Statement.php]: Hmm.. okay. `DriverInterface` is quite redundant too, no one's going to implement their own implementation of it. `Driver` class will be always used.

File: src/Database/Statement.php (hunk 7 above)
Comment (ADmad) [reply src/Database/Statement.php]: Yes, the currently code is quite clunky.

File: src/Database/Statement.php (hunk 9 above)
Comment (ADmad) [reply src/Database/Statement.php]: Well `rowCount()` does work for all (most) dbs besides Sqlite even for SELECT and is shown in DebugKit currently.

File: src/Datasource/FixtureInterface.php [hunk 10]
```diff
@@ -16,7 +16,7 @@
  */
 namespace Cake\Datasource;
 
-use Cake\Database\StatementInterface;
+use Cake\Database\Statement;
```
Comment (ADmad) [inline src/Datasource/FixtureInterface.php]: This is a separate issue but datasource package shouldn't have a dependency on a database package class :disappointed:

File: src/ORM/Query.php (hunk 1 above)
Comment (ADmad) [reply src/ORM/Query.php]: We can look into changing the arguments which accept this resultset to `iterable` instead of `array`.

File: src/ORM/Table.php [hunk 11]
```diff
@@ -2105,9 +2105,14 @@ protected function _insert(EntityInterface $entity, array $data): EntityInterfac
             $entity->set($filteredKeys, ['guard' => false]);
             $schema = $this->getSchema();
             $driver = $this->getConnection()->getDriver();
+            $returning = $statement->fetchAssoc();
```
Comment (ADmad) [inline src/ORM/Table.php]: This is better than using `PDO::lastInsertId()`?

Comment (othercorey) [reply src/ORM/Table.php]: For drivers that support RETURNING and similar syntax, the query returns a row with the generated IDs. PDO::lastInsertId() either returns the last ID for a table or for a sequence. For example, you can't call lastInsertId() for a table on postgresql. You must fetch the row retuned from INSERT.

File: src/Database/Connection.php [hunk 12]
```diff
@@ -448,9 +446,9 @@ public function update(string $table, array $values, array $conditions = [], arr
      * @param string $table the table to delete rows from
      * @param array $conditions conditions to be set for delete statement
      * @param array $types list of associative array containing the types to be used for casting
-     * @return \Cake\Database\StatementInterface
+     * @return \Cake\Database\Statement
```
Comment (markstory) [inline src/Database/Connection.php]: Why don't we want to return an interface here?

File: src/Database/Query.php [hunk 13]
```diff
@@ -149,11 +149,16 @@ class Query implements ExpressionInterface, IteratorAggregate, Stringable
...
+    protected ?Statement $_statement = null;
+
+    /**
+     * Result set from exeuted SELCT query.
      *
-     * @var \Cake\Database\StatementInterface|null
+     * @var iterable|null
      */
-    protected ?StatementInterface $_iterator = null;
+    protected ?iterable $_results = null;
```
Comment (markstory) [inline src/Database/Query.php:181]: If we're adding new properties could we skip the `_` prefix?

File: src/Database/Query.php [hunk 14]
```diff
@@ -237,15 +242,40 @@ public function getConnection(): Connection
...
-        $this->_iterator = $this->_decorateStatement($statement);
+        $this->_statement = $this->_connection->run($this);
         $this->_dirty = false;
 
-        return $this->_iterator;
+        return $this->_statement;
+    }
+
+    /**
+     * Executes query and returns entire result set.
```
Comment (markstory) [inline src/Database/Query.php]: This reads like the output will be a ResultSet.
```suggestion
     * Executes query and returns the decorated results.
```

File: src/Database/Statement.php (hunk 3 above)
Comment (markstory) [reply src/Database/Statement.php]: If we have no scenario where we expect users to swap in their own statements then we could do without the interfaces. It does force us to use inheritance in our other statements instead of decoration as well.

File: src/Database/Statement.php (hunk 7 above)
Comment (markstory) [reply src/Database/Statement.php]: Could we take a similar approach to how we do error logging where the metadata (duration, row counts) is sent to a logger that then decides on whether or not to log the query? That might be a breaking change but again I suspect there are very few database logging implementations in the wild so the impact would be low.

File: src/Database/Connection.php (hunk 12 above)
Comment (othercorey) [reply src/Database/Connection.php]: There's no reason to support an Interface now that we don't use the nested decorator pattern. We expect a `Statement`.

File: src/Database/Query.php (hunk 13 above)
Comment (othercorey) [reply src/Database/Query.php:181]: This is caused by `QueryTrait`. ORM\Query doesn't actually implement its own features. It uses `QueryTrait` which defines `$_results`. 

It's awkward to have a trait overrides the base class, but it seemed better to line it up.

File: src/Database/Statement.php (hunk 3 above)
Comment (othercorey) [reply src/Database/Statement.php]: I updated the design to support inheritance in driver-specific statements better. Notice how `SqlserverStatement` doesn't have to duplicate the bind logic anymore.

File: src/Database/Statement.php (hunk 7 above)
Comment (ADmad) [reply src/Database/Statement.php]: Not sure if that would be the right approach. Logging is an optional feature. So if it's not enabled then a logger instance wouldn't/shouldn't exist.

File: src/Database/Driver.php (hunk 4 above)
Comment (othercorey) [reply src/Database/Driver.php]: Whoops, I misread this. Will have to add new constants.

File: src/Database/Driver/Sqlserver.php (hunk 5 above)
Comment (othercorey) [reply src/Database/Driver/Sqlserver.php]: Fixed.

File: src/Database/Statement.php (hunk 8 above)
Comment (othercorey) [reply src/Database/Statement.php]: Fixed.

File: src/Database/Statement.php (hunk 9 above)
Comment (othercorey) [reply src/Database/Statement.php]: Fixed.

File: src/ORM/Table.php (hunk 11 above)
Comment (othercorey) [reply src/ORM/Table.php]: Reverted.

File: src/Database/Statement.php (hunk 3 above)
Comment (othercorey) [reply src/Database/Statement.php]: Re-introduced StatementInterface.

File: src/Database/Connection.php (hunk 12 above)
Comment (othercorey) [reply src/Database/Connection.php]: Re-introduced StatementInterface.

File: src/Database/Connection.php [hunk 15]
```diff
@@ -280,9 +279,8 @@ public function prepare(Query|string $query): StatementInterface
     {
         return $this->getDisconnectRetry()->run(function () use ($query) {
             $statement = $this->_driver->prepare($query);
-
             if ($this->_logQueries) {
-                $statement = $this->_newLogger($statement);
+                $statement->setLogger($this->getLogger());
```
Comment (othercorey) [inline src/Database/Connection.php:283]: Should we make `setLogger()` part of `StatementInterface`? Otherwise, we have to type that this is a `Statement` instance.

Comment (ADmad) [reply src/Database/Connection.php:283]: Could we instead come up with a solution which doesn't require passing the logger into the statement? The driver could do the logging after it executes the statement.

File: src/ORM/Query.php (hunk 1 above)
Comment (othercorey) [reply src/ORM/Query.php]: Updated to convert to `array` before passing to `loadExternal()`. Will work on results sets separately.

File: src/Database/Driver.php (hunk 4 above)
Comment (othercorey) [reply src/Database/Driver.php]: Added TYPE_SELECT constant to Query.

File: src/Database/Statement.php (hunk 7 above)
Comment (othercorey) [reply src/Database/Statement.php]: Cleaned up logging in execute(). Only runs if logger is set.

File: src/Database/Connection.php (hunk 15 above)
Comment (othercorey) [reply src/Database/Connection.php:283]: Added to StatementInterface for now. Will revisit separately.

**Comment by othercorey** [comment]
Update:

* Re-introduced `StatementInterface`. We can revisit that later.
* Moved `Statement` to `Database\Statement\Statement`
* Changed `ORM\Query::all()` to convert iterable to array before passing to `EagerLoader`
* Added `Query::TYPE_*` constants to replace `'select'`
* Added Statement::queryString()
* Cleaned up query logging
* Added `StatementTest` which ports the test from `LoggingStatementTest`

File: src/Database/Statement/Statement.php [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+
+    /**
+     * @var \Cake\Database\DriverInterface
+     */
+    protected DriverInterface $_driver;
+
+    /**
+     * @var \PDOStatement
+     */
+    protected PDOStatement $statement;
```
Comment (ADmad) [inline src/Database/Statement/Statement.php:55]: I would rename this to `$pdoStatement` for better readability.

Comment (othercorey) [reply src/Database/Statement/Statement.php:55]: I can do that.

File: src/Database/Connection.php (hunk 15 above)
Comment (markstory) [reply src/Database/Connection.php:283]: Since statements already have access to the driver, could we get the logger from the driver? If logging is disabled we could use a NullObject to do a no-op.

File: src/Database/Query.php (hunk 2 above)
Comment (markstory) [reply src/Database/Query.php:274]: Could we raise an error if `_type !== Query::TYPE_SELECT && $this->_resultDecorators`?

File: src/Database/Statement/Statement.php [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+        $this->statement = $statement;
+        $this->typeConverter = $typeMap !== null ? new FieldTypeConverter($typeMap, $driver) : null;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function setLogger(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
```
Comment (markstory) [inline src/Database/Statement/Statement.php:99]: Instead of setting the logger on each statement, could we get the logger from the driver? It would require pushing the logger down from `Connection` into the `Driver` but it would remove some complexity here.

If you'd like I can take a swing at doing that change.

File: src/Database/Query.php (hunk 2 above)
Comment (othercorey) [reply src/Database/Query.php:274]: I call execute() from all(). I'd have to move this to a protected method.

File: src/Database/Statement/Statement.php (hunk 17 above)
Comment (othercorey) [reply src/Database/Statement/Statement.php:99]: What is the difference between Connection and Driver long-term? Do we want Driver to maintain state?

Comment (markstory) [reply src/Database/Statement/Statement.php:99]: Long term I don't know. The entire Datasources namespace is a bit suspect and I'm not sure it is justified if we are going to focus on the just the ORM and make plugins like Elasticsearch totally separate but 'ORM like'.
